clients/pkg/promtail/targets/file/filetarget_test.go

```diff
@@ -221,6 +232,101 @@ func TestFileTarget_StopsTailersCleanly(t *testing.T) {
 
 	target.Stop()
 	ps.Stop()
+
+	require.NoError(t, testutil.GatherAndCompare(registry, bytes.NewBufferString(`
+		# HELP promtail_files_active_total Number of active files.
+		# TYPE promtail_files_active_total gauge
+		promtail_files_active_total 0
+	`), "promtail_files_active_total"))
+}
+
+func TestFileTarget_StopsTailersCleanly_Parallel(t *testing.T) {
+	w := log.NewSyncWriter(os.Stderr)
+	logger := log.NewLogfmtLogger(w)
+
+	tempDir := t.TempDir()
+	positionsFileName := filepath.Join(tempDir, "positions.yml")
+
+	ps, err := positions.New(logger, positions.Config{
+		SyncPeriod:    10 * time.Millisecond,
+		PositionsFile: positionsFileName,
+	})
+	require.NoError(t, err)
+
+	client := fake.New(func() {})
+	defer client.Stop()
+
+	pathToWatch := filepath.Join(tempDir, "*.log")
+	registry := prometheus.NewRegistry()
+	metrics := NewMetrics(registry)
+
+	// Increase this to several thousand to make the test more likely to fail when debugging a race condition
+	iterations := 500
+	fakeHandler := make(chan fileTargetEvent, 10*iterations)
+	for i := 0; i < iterations; i++ {
```

Comment:
This test will take about 10s and has a good chance of catching the issue, but to catch this problem reliably I had to run 5k+ iterations which takes 2-3min.